### PR TITLE
crypto: ensure domain separator between address and object id

### PR DESCRIPTION
**Pull request by joyqvq**
## Description 

Closes https://github.com/MystenLabs/sui/issues/9544
## Test Plan 

Existing tests
---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-adapter-transactional-tests/tests/shared/upgrade.exp [hunk 1]
```diff
@@ -12,15 +12,18 @@ created: object(107), object(108), object(109), object(110)
...
 
 task 3 'view-object'. lines 44-44:
-Owner: Object ID: ( fake(107) )
+Owner: Account Address ( A )
 Version: 2
 Contents: t::m::Obj {id: sui::object::UID {id: sui::object::ID {bytes: fake(110)}}}
 
 task 4 'run'. lines 46-46:
-Error: Object 0xe6ff026a903dda0711f951bab1d01161791f0f6fe94badadcb0424cf05bd591d is owned by object 0xf2c06d97f4ad9da4a8a6cfb80dd73173302be0c6c245c62279a4bc86a366f33c. Objects owned by other objects cannot be used as input arguments.
+Error: Transaction Effects Status: Move Runtime Abort. Location: sui::transfer::share_object_impl (function index 7) at offset 0, Abort Code: 0
```
Comment (kchalkias) [inline crates/sui-adapter-transactional-tests/tests/shared/upgrade.exp:20]: Are these expected? I see some errors here

Comment (tnowacki) [reply crates/sui-adapter-transactional-tests/tests/shared/upgrade.exp:20]: It is likely that the fake IDs get shuffled around.... I really which I had a different way of doing this, but the  IDs get sorted, so anytime the generation gets changed (which shouldn't happen often) these tests break

Comment (joyqvq) [reply crates/sui-adapter-transactional-tests/tests/shared/upgrade.exp:20]: should i change anything to the tests?

File: crates/shared-crypto/src/intent.rs [hunk 2]
```diff
@@ -170,10 +170,12 @@ pub(crate) mod private {
     impl<T> SealedIntent for IntentMessage<T> {}
 }
 
-/// Hashing intent is used to prevent hash collision when hashing a message.
+/// A 1-byte domain separator for hashing Object ID in Sui. It is starting from 0xf0
+/// to ensure no hashing collision for any ObjectID vs SuiAddress which is derived
+/// as the hash of `flag || pubkey`. See [`sui_types::crypto::SignatureScheme::flag()`].
```
Comment (kchalkias) [inline crates/shared-crypto/src/intent.rs:175]: we can avoid the link here (add it as simple string so CI doesn't complain), and we can fix cross-dep doc links in upcoming DOCathons.

**Review by kchalkias — COMMENTED** [review]
@tnowacki I don't understand what the next step is here. Do we need to change anything for these tests not to fail?

**Comment by tnowacki** [comment]
> @tnowacki I don't understand what the next step is here. Do we need to change anything for these tests not to fail?

I fixed them :)

File: crates/sui-adapter-transactional-tests/tests/child_count/non_zero_child_count_valid.exp [hunk 3]
```diff
@@ -21,21 +21,21 @@ created: object(112), object(113), object(114)
 written: object(111)
 
 task 5 'run'. lines 58-58:
-written: object(113), object(115)
+Error: Object 0x4ed62609e80f1a882d16311822ec560fbc1d81a2ab0bcb4fd6358910dee9573f is owned by object 0x0e1963f04364076254e2239b46ad24cccf0ce9c4adec48c942e5180ff59ed753. Objects owned by other objects cannot be used as input arguments.
```
Comment (kchalkias) [inline crates/sui-adapter-transactional-tests/tests/child_count/non_zero_child_count_valid.exp:24]: @tnowacki, so is this expected now for the behavior to differ vs before that PR?

File: crates/sui-adapter-transactional-tests/tests/dynamic_fields/wrap_object.exp [hunk 4]
```diff
@@ -37,13 +37,14 @@ created: object(117), object(118), object(119)
...
-Owner: Object ID: ( fake(118) )
+Owner: Object ID: ( fake(117) )
 Version: 2
-Contents: sui::dynamic_field::Field<sui::dynamic_object_field::Wrapper<u64>, sui::object::ID> {id: sui::object::UID {id: sui::object::ID {bytes: fake(119)}}, name: sui::dynamic_object_field::Wrapper<u64> {name: 0u64}, value: sui::object::ID {bytes: fake(117)}}
+Contents: sui::dynamic_field::Field<sui::dynamic_object_field::Wrapper<u64>, sui::object::ID> {id: sui::object::UID {id: sui::object::ID {bytes: fake(119)}}, name: sui::dynamic_object_field::Wrapper<u64> {name: 0u64}, value: sui::object::ID {bytes: fake(118)}}
 
 task 10 'run'. lines 61-61:
-written: object(117), object(118), object(120)
-deleted: object(119)
+Error: Object 0xeee3950b4520d1b95dc25fd3bc0c50aa118d10369eec4a2d60476da5c1309f66 is owned by object 0x7ac9fc785d7ee8edae992ef836b7a02a7d2d132bc5756a7948f6fef0ca5f3203. Objects owned by other objects cannot be used as input arguments.
```
Comment (kchalkias) [inline crates/sui-adapter-transactional-tests/tests/dynamic_fields/wrap_object.exp:45]: here as well @tnowacki. Iirc when Todd pushed his code these were resolved, but after @joyqvq's latest push they are inconsistent again? (I'm not sure actually, I just have the impression I've seen a state where they did indeed match).

File: crates/sui-adapter-transactional-tests/tests/shared/upgrade.exp (hunk 1 above)
Comment (kchalkias) [reply crates/sui-adapter-transactional-tests/tests/shared/upgrade.exp:20]: thanks @tnowacki, I guess you refer to the whole [upgrade.exp] file. Then @joyqvq we're good to go

**Review by kchalkias — APPROVED** [review]

**Pull request merged**